server/src/main/java/org/opensearch/identity/ScopeAwareSubject.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity;
+
+import java.util.Set;
+import org.opensearch.identity.scopes.Scope;
+
+/**
+ * This interface defines a ScopeAwareSubject.
+ *
+ * A ScopeAwareSubject is an extension of the Subject interface which adds the concept of scopes.
+ * Specifically, a ScopeAwareSubject must be able to get its own scopes (as a set of Strings) and set its own scopes,
+ *
+ */
+public interface ScopeAwareSubject extends Subject {
+
+    Set<String> getScopes();
+
+    void setScopes(Set<Scope> scopes);
```

Comment:
This seems messier than ApplicationAwareSubject, writes to `setScopes` need to be saved to data store that could be external to persist the change - is this what you had in mind?

Reply:
I made this swap because I felt that the old setup of using ApplicationAwareSubject etc. limited what we could do with scopes. If we used the old implementation, we had no way to persists the changes as you mentioned. Likewise, it did not seem appropriate to reference user/client objects using that class. Because we lacked an interface, we would not be able to implement later setups such as a ScopedInternalUser which could be a form of access control native to a identity system living exclusively in core. With a ScopeAwareSubject interface, we are able to make the implementations generic and then can use the different types of scopes for different use cases as they arise.